federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,827 @@
+		return false
+	}
+
+	if x.Port != y.Port {
+		return false
+	}
+
+	if x.NodePort != y.NodePort {
+		return false
+	}
+
+	// We don't check TargetPort; that is not relevant for load balancing
+	// TODO: Should we blank it out?  Or just check it anyway?
+
+	return true
+}
+
+func intSlicesEqual(x, y []int) bool { // TODO: quinton - this belongs in the utils library, not here.  Move it.
+	if len(x) != len(y) {
+		return false
+	}
+	if !sort.IntsAreSorted(x) {
+		sort.Ints(x)
+	}
+	if !sort.IntsAreSorted(y) {
+		sort.Ints(y)
+	}
+	for i := range x {
+		if x[i] != y[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func stringSlicesEqual(x, y []string) bool { // TODO: quinton - this belongs in the utils library, not here.  Move it.
+	if len(x) != len(y) {
+		return false
+	}
+	if !sort.StringsAreSorted(x) {
+		sort.Strings(x)
+	}
+	if !sort.StringsAreSorted(y) {
+		sort.Strings(y)
+	}
+	for i := range x {
+		if x[i] != y[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func hostsFromServiceList(list *api.ServiceList) []string { // TODO: quinton: This probably needs to return IP's or DNS names, not "hostnames".  Rename and recode.
+	result := []string{}
+	for ix := range list.Items {
+		if list.Items[ix].Spec.Unschedulable { // TODO: quinton: In the case of cluster services this relates to the health of the services, not the nodes.  Fix this.
+			continue
+		}
+		result = append(result, list.Items[ix].Name)
+	}
+	return result
+}
+
+//TODO: quinton: Add a good comment/doc
+func getClusterServiceConditionPredicate() cache.ClusterServiceConditionPredicate {
+	return func(clusterService kube_api.Service) bool {
+		if clusterService.Status.Health != OK { //TODO: quinton: Fix the health check logic here.
+			return false
+		}
+		// If we have no info, don't accept
+		if len(clusterService.Status.Conditions) == 0 { // TODO: quinton: Check whether this applies
+			return false
+		}
+		for _, cond := range clusterService.Status.Conditions {
+			// We consider the cluster service for inclusion in DNS records only when its ServiceReady condition status
+			// is ConditionTrue
+			// TODO: quinton: Check up on exactly how service status works, and fix.
+			if cond.Type == api.ServiceReady && cond.Status != api.ConditionTrue {
+				glog.V(4).Infof("Ignoring cluster service %v with %v condition status %v", clusterService.Name, cond.Type, cond.Status)
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// clusterServiceSyncLoop handles updating the load balancers pointed to by all federated service DNS Records
+// whenever the set of cluster services in the federation changes.
+// TODO: quinton: Assumed that cluster service load balancers have host (i.e. DNS) names (and IP's?).  Verify and fix as necessary.
+//                Also the logic here bundles all hos/DNS names of all services in the federation together - need to handle service one separately.  Rewrite.
+func (s *ServiceController) clusterServiceSyncLoop(period time.Duration) {
```

Comment:
@mfanjie Right now the dnsprovider interface accepts a single hostname, but I plan to add the multi-hostname function also, as you suggest.  The API will look almost identical to the above.